dm/master/openapi.go

```diff
@@ -133,17 +133,102 @@ func (s *Server) DMAPIDeleteSource(ctx echo.Context, sourceName string) error {
 
 // DMAPIStartRelay url is:(POST /api/v1/sources/{source-id}/relay).
 func (s *Server) DMAPIStartRelay(ctx echo.Context, sourceName string) error {
-	return nil
+	var req openapi.StartRelayRequest
+	if err := ctx.Bind(&req); err != nil {
+		return err
+	}
+	sourceCfg := s.scheduler.GetSourceCfgByID(sourceName)
+	if sourceCfg == nil {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
+	}
+	// only need to update source config when relay log related config is changed
+	needUpdate := false
+	if req.RelayBinlogName != nil && sourceCfg.RelayBinLogName != *req.RelayBinlogName {
+		sourceCfg.RelayBinLogName = *req.RelayBinlogName
+		needUpdate = true
+	}
+	if req.RelayBinlogGtid != nil && sourceCfg.RelayBinlogGTID != *req.RelayBinlogGtid {
+		sourceCfg.RelayBinlogGTID = *req.RelayBinlogGtid
+		needUpdate = true
+	}
+	if req.RelayDir != nil && sourceCfg.RelayDir != *req.RelayDir {
+		sourceCfg.RelayDir = *req.RelayDir
+		needUpdate = true
+	}
+	if purge := req.Purge; purge != nil {
+		if purge.Expires != nil && sourceCfg.Purge.Expires != *purge.Expires {
+			sourceCfg.Purge.Expires = *purge.Expires
+			needUpdate = true
+		}
+		if purge.Interval != nil && sourceCfg.Purge.Interval != *purge.Interval {
+			sourceCfg.Purge.Interval = *purge.Interval
+			needUpdate = true
+		}
+		if purge.RemainSpace != nil && sourceCfg.Purge.RemainSpace != *purge.RemainSpace {
+			sourceCfg.Purge.RemainSpace = *purge.RemainSpace
+			needUpdate = true
+		}
+	}
+	if needUpdate {
+		// update current source relay config before start relay
+		if err := s.scheduler.UpdateSourceCfg(sourceCfg); err != nil {
+			return err
+		}
+	}
+	return s.scheduler.StartRelay(sourceName, []string{req.WorkerName})
 }
 
 // DMAPIStopRelay url is:(DELETE /api/v1/sources/{source-id}/relay).
 func (s *Server) DMAPIStopRelay(ctx echo.Context, sourceName string) error {
-	return nil
+	var req openapi.WorkerNameRequest
+	if err := ctx.Bind(&req); err != nil {
+		return err
+	}
+	return s.scheduler.StopRelay(sourceName, []string{req.WorkerName})
 }
 
 // DMAPIGetSourceStatus url is:(GET /api/v1/sources/{source-id}/status).
 func (s *Server) DMAPIGetSourceStatus(ctx echo.Context, sourceName string) error {
-	return nil
+	sourceCfg := s.scheduler.GetSourceCfgByID(sourceName)
+	if sourceCfg == nil {
+		return terror.ErrSchedulerSourceCfgNotExist.Generate(sourceName)
+	}
+	workerName := ""
+	if w := s.scheduler.GetWorkerBySource(sourceName); w != nil {
+		workerName = w.BaseInfo().Name
+	}
+	resp := openapi.SourceStatus{
+		EnableRelay: sourceCfg.EnableRelay,
+		SourceName:  sourceCfg.SourceID,
+		WorkerName:  workerName,
+	}
+	// only query worker if source is bound to a worker
+	queryWorker := workerName != ""
+	if queryWorker {
+		ret := s.getStatusFromWorkers(ctx.Request().Context(), []string{sourceName}, "", sourceCfg.EnableRelay)
+		if len(ret) != 1 {
```

Comment:
many workers can handle one source's relay

Reply:
I think the api should reflect the current status of the source , so only the stauts of the bound source worker will be shown here, when one source was transfer to worker2, we only show worker2's relay status🤔

Reply:
is there other API that shows status of multiple "relay workers"?

Reply:
currently no, maybe we can add this api when we support one worker handle mutiple source

Reply:
currenly we did support one source has multiple relay workers. If some relay worker has error, we have no API to know it. But dmctl can use `query-status -s <source-id>` to show all relay workers' status.

Reply:
we can add another api to show all source status in next part of OpenAPI,  so this api only shows the status of the currently bound worker